### LATAM CHALLENGE - Yohan Sebastian 
Este challenge consistió en realizar optimizaciones en cuanto a rendimiento y memoria en diversos casos de uso

### Resumen

Los datos analizados fueron proporcionados por LATAM, los cuales consisten en un archivo JSON de 400Mb de tamaño. 
El desarrollo del reto propuesto será desarrollado en un notebook de python utilizando un editor de código tradiciones como lo es VS Code, el repositorio de git en el cual se alojará la solución está en mi cuenta personal de GitHub, no obstante se hará público para su revisión. 
El código adjunto puede ser fácilmente llevado a la nube, bien sea de google - aws - azure u otra, eligiendo el mejor servicio para su despliegue, se recomiendan usar tecnologías serverless para su ejecución ya que optimizan rendimiento y recursos.

Se importan los módulos correspondientes desde src, de modo que podemos acceder a los diferentes métodos desarrollados en el presente challenge. Las librerías usadas a instalar se relacionan en el archivo de requirements.txt, la versión de python utilizada es la 3.8

In [34]:
from src import q1_memory, q1_time, q2_memory, q2_time, q3_memory, q3_time 

El file_path que relaciono en la siguiente celda es un file_path de mi computadora local personal, sin embargo dicho path corresponde a la ruta donde se aloja la data proporcionada por LATAM. Si el proyecto se lleva a la nube, dicho path correspondería a la ruta del sistema de almacenamiento de preferencia, por ejemplo en google cloud storage, aws S3, azure datalake.

In [15]:
#definimos el path json de la data y los comandos magic para evaluar tiempo y memoria
file_path = "C:/Users/User/Desktop/arquitectura/tweets.json/farmers-protest-tweets-2021-2-4.json"
%load_ext memory_profiler
%load_ext line_profiler

The memory_profiler extension is already loaded. To reload it, use:
  %reload_ext memory_profiler
The line_profiler extension is already loaded. To reload it, use:
  %reload_ext line_profiler


A continuación se llamarán los métodos desarrollados en los diferentes módulos, y a cada uno de ellos se les aplicarán las funciones que describen el comportamiento en cuanto a memoria y rendimiento

In [4]:
q1_time.q1_time(file_path)

[(datetime.date(2021, 2, 12), 'RanbirS00614606', 176),
 (datetime.date(2021, 2, 13), 'MaanDee08215437', 178),
 (datetime.date(2021, 2, 14), 'rebelpacifist', 119),
 (datetime.date(2021, 2, 15), 'jot__b', 134),
 (datetime.date(2021, 2, 16), 'jot__b', 133),
 (datetime.date(2021, 2, 17), 'RaaJVinderkaur', 185),
 (datetime.date(2021, 2, 18), 'neetuanjle_nitu', 195),
 (datetime.date(2021, 2, 19), 'Preetm91', 267),
 (datetime.date(2021, 2, 20), 'MangalJ23056160', 108),
 (datetime.date(2021, 2, 23), 'Surrypuria', 135)]

In [18]:
%timeit q1_time.q1_time(file_path)

7.69 s ± 1.04 s per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [19]:
%memit q1_time.q1_time(file_path)

peak memory: 549.52 MiB, increment: 430.46 MiB


In [16]:
q1_memory.q1_memory(file_path)

[('2021-02-12', 'RanbirS00614606'),
 ('2021-02-13', 'MaanDee08215437'),
 ('2021-02-17', 'RaaJVinderkaur'),
 ('2021-02-16', 'jot__b'),
 ('2021-02-14', 'rebelpacifist'),
 ('2021-02-18', 'neetuanjle_nitu'),
 ('2021-02-15', 'jot__b'),
 ('2021-02-20', 'MangalJ23056160'),
 ('2021-02-23', 'Surrypuria'),
 ('2021-02-19', 'Preetm91')]

In [20]:
%timeit q1_memory.q1_memory(file_path)

32.6 s ± 876 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [21]:
%memit q1_memory.q1_memory(file_path)

peak memory: 119.11 MiB, increment: 0.00 MiB


In [22]:
q2_time.q2_time(file_path)

[('🙏', 5049),
 ('😂', 3072),
 ('🚜', 2972),
 ('🌾', 2182),
 ('🇮🇳', 2086),
 ('🤣', 1668),
 ('✊', 1651),
 ('❤️', 1382),
 ('🙏🏻', 1317),
 ('💚', 1040)]

In [23]:
%timeit q2_time.q2_time(file_path)

20.9 s ± 1.19 s per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [24]:
%memit q2_time.q2_time(file_path)

peak memory: 136.84 MiB, increment: 29.46 MiB


In [25]:
q2_memory.q2_memory(file_path)

[('🙏', 5049),
 ('😂', 3072),
 ('🚜', 2972),
 ('🌾', 2182),
 ('🇮🇳', 2086),
 ('🤣', 1668),
 ('✊', 1651),
 ('❤️', 1382),
 ('🙏🏻', 1317),
 ('💚', 1040)]

In [26]:
%timeit q2_memory.q2_memory(file_path)

22.8 s ± 1.69 s per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [27]:
%memit q2_memory.q2_memory(file_path)

peak memory: 117.89 MiB, increment: 0.00 MiB


In [38]:
q3_time.q3_time(file_path)

[('narendramodi', 2265),
 ('Kisanektamorcha', 1840),
 ('RakeshTikaitBKU', 1644),
 ('PMOIndia', 1427),
 ('RahulGandhi', 1146),
 ('GretaThunberg', 1048),
 ('RaviSinghKA', 1019),
 ('rihanna', 986),
 ('UNHumanRights', 962),
 ('meenaharris', 926)]

In [40]:
%timeit q3_time.q3_time(file_path)

5.09 s ± 52.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [44]:
%memit q3_time.q3_time(file_path)

peak memory: 90.34 MiB, increment: 0.00 MiB


In [41]:
q3_memory.q3_memory(file_path)

[('narendramodi', 2265),
 ('Kisanektamorcha', 1840),
 ('RakeshTikaitBKU', 1644),
 ('PMOIndia', 1427),
 ('RahulGandhi', 1146),
 ('GretaThunberg', 1048),
 ('RaviSinghKA', 1019),
 ('rihanna', 986),
 ('UNHumanRights', 962),
 ('meenaharris', 926)]

In [43]:
%timeit q3_memory.q3_memory(file_path)

5.4 s ± 346 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [42]:
%memit q3_memory.q3_memory(file_path)

peak memory: 90.31 MiB, increment: 0.00 MiB


### CONCLUSIONES
Comparaciones finales de rendimeinto y memoria para cada uno de los casos estudiados


| Funcion | Tiempo | Memoria |
|---------|--------|---------|
| q1_time | 7.69 | 549.52 Mib |
| q1_memory | 32.6s  | 119.11 MiB |
| q2_time | 20.9 | 136.84 Mib |
| q2_memory | 22.8s  | 117.89 MiB |
| q3_time | 5.09s  | 90.34 MiB |
| q3_memory | 5.4 | 90.31 Mib |

Consideraciones:
El particionamiento de los datos utilizado en el q1_memory mostró una reducción muy significativa en cuanto al uso de memoria, sin embargo se sacrificó rendimiento, dicho rendimiento pudo maximizarse si hubiesemos ejecutado las tareas en hilos o en un cluster donde múltiples workers realizarán las diferentes operaciones.
La lectura y procesamiento línea a línea mostró mejor desempeño en memoria que la lectura completa del archivo, sin embargo en términos de tiempo/rendimiento fue mejor consolidar el archivo para su posterior procesamiento.
El manejo de condicionales y filtros hacen ligeramente en el caso q3_time más eficiente el código en rendimiento, sin embargo en cuanto a memoria no se mostró mayor variación entre un caso y el otro.
